### PR TITLE
Ported matmul compositeimplicitautograd impl into core

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85239

### CODE REVIEW

**Comment by Chillee** [comment]
Still a bit WIP

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
looks like you need reshape

**Comment by Chillee** [comment]
What's up with reshape? Have we just not ported it from the branch?

**Comment by ezyang** [comment]
nope, reshape is not in master yet

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -3516,14 +3516,16 @@ static void linalg_lstsq_out_info(
       at::sum_out(residuals, raw_residuals, /*dim=*/-2, /*keepdim=*/false, /*dtype*/real_dtype);
     }
   }
-  solution = solution.narrow(/*dim=*/-2, /*start=*/0, /*length*/n);
+  auto solution_view = solution.narrow(/*dim=*/-2, /*start=*/0, /*length*/n);
+  // manually restride original
+  solution.set_(solution.storage(), solution_view.storage_offset(), solution_view.sizes(), solution_view.strides());
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: What has changed with `narrow` and why is this change needed?

**Comment by Chillee** [comment]
🤔 I somehow accidentally merged https://github.com/pytorch/pytorch/pull/85322 into this PR. Since @ezyang also approved that PR, I'll just merge this PR :^)

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (Chillee) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: @ezyang can comment on this more tomorrow, but to quote: "writing over Tensor& is not kosher".

This was causing some double-free errors elsewhere.

Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: Yes, writing over Tensor&, in general, should be avoided.
Why does it have to be in this PR?

Comment (Chillee) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: Enabling `enable_python_dispatcher` in `test/test_decomp.py` triggered errors with this code and corrupted memory.

image

https://github.com/pytorch/pytorch/actions/runs/3085755988/jobs/4989537077

Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: Why do you peek inside the `CompositeExplicitAutograd` function at all?
Maybe correct fix would be to change `CompositeExplicitAutograd: linalg_lstsq -> CPU, CUDA: linalg_lstsq` here: https://github.com/pytorch/pytorch/blob/bcdef58a55665bba959a773633aa5b1e5758d94e/aten/src/ATen/native/native_functions.yaml#L12246 ?

Comment (Chillee) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: We're not peeking inside `CompositeExplicitAutograd`. `enable_python_dispatcher` allows us to register python implementations of `CompositeImplicitAutograd` functions (like you've been asking for!), and this PR adds a python implementation of `aten::matmul`. `linalg_lstsq` backward formula uses a matmul, and so we enter the decomposition for it (where we see the shape error).

Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: > like you've been asking for

I don't remember asking for it 😅 

There was no shape error before this PR, could it mean that Python's implementation in this PR is different from `aten::matmul`?

Comment (Chillee) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: Haha, I just mean that this allows us to make use of `CompositeImplicitAutograd` decompositions, while previously they weren't so useful :P 

> could it mean that Python's implementation in this PR is different from aten::matmul

Yeah, that's a reasonable thing to guess. But in this case, the test was failing from `enable_python_dispatcher` even without adding the `matmul` decomp :P

**Review by IvanYashchuk — COMMENTED** [review]
Here's the problem (changes to `aten/src/ATen/native/BatchLinearAlgebra.cpp` are reverted):
```py
In [1]: import torch

In [2]: a = torch.randn(4, 3, device="cuda")

In [3]: b = torch.rand(4, 3, device="cuda")

In [4]: expected_shape = torch.linalg.lstsq(a, b).solution.shape

In [5]: from torch._dispatch.python import enable_python_dispatcher

In [6]: with enable_python_dispatcher():
   ...:     python_disp_shape = torch.linalg.lstsq(a, b).solution.shape
   ...:

In [7]: expected_shape == python_disp_shape
Out[7]: False

In [8]: expected_shape, python_disp_shape
Out[8]: (torch.Size([3, 3]), torch.Size([4, 3]))
```
What's causing `solution.shape` to be different under `enable_python_dispatcher()`?

**Review by lezcano — APPROVED** [review]
The matmul / mm / etc changes look good to me

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: Could you please revert this?

**Review by IvanYashchuk — CHANGES_REQUESTED** [review]
With `enable_python_dispatcher` for some reason `at::linalg_lstsq_outf` returns a different result than `at::native::linalg_lstsq_out`.

Instead of introducing changes to `linalg_lstsq_out_info` helper function. Could you please apply the following diff, it changes `at::linalg_lstsq_outf -> at::native::linalg_lstsq_out`:
```
diff --git a/aten/src/ATen/native/BatchLinearAlgebra.cpp b/aten/src/ATen/native/BatchLinearAlgebra.cpp
index f74a282194..aec4e2c3c3 100644
--- a/aten/src/ATen/native/BatchLinearAlgebra.cpp
+++ b/aten/src/ATen/native/BatchLinearAlgebra.cpp
@@ -3516,16 +3516,14 @@ static void linalg_lstsq_out_info(
       at::sum_out(residuals, raw_residuals, /*dim=*/-2, /*keepdim=*/false, /*dtype*/real_dtype);
     }
   }
-  auto solution_view = solution.narrow(/*dim=*/-2, /*start=*/0, /*length*/n);
-  // manually restride original
-  solution.set_(solution.storage(), solution_view.storage_offset(), solution_view.sizes(), solution_view.strides());
+  solution = solution.narrow(/*dim=*/-2, /*start=*/0, /*length*/n);
   if (m == 0) {
     solution.zero_();
   }

   // for 1-dimensional 'other', we need to squeeze the solution after "apply_lstsq"
   if (vector_case) {
-    solution.squeeze_(-1);
+    solution = solution.squeeze_(-1);
   }
 }

@@ -3735,7 +3733,7 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> linalg_lstsq(
   Tensor rank = at::empty({0}, input.options().dtype(at::kLong));
   Tensor singular_values = at::empty({0}, input.options().dtype(toRealValueType(input.scalar_type())));
   std::tie(solution, residuals, rank, singular_values) =
-      at::linalg_lstsq_outf(input, other, rcond, driver, solution, residuals, rank, singular_values);
+      at::native::linalg_lstsq_out(input, other, rcond, driver, solution, residuals, rank, singular_values);
   return std::make_tuple(solution, residuals, rank, singular_values);
 }
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: It's a memory error, caused by the overwriting of the mutable Tensor&. Once you have a memory error, all bets are off; in this case, it manifests as the shape mismatched, but the true problem is the memory error. You can determine this by running the program with ASAN with the python dispatcher enabled.

The reason the python dispatcher matters is because it causes all dispatched calls (that includes the internal `at::squeeze`, etc. calls inside the CompositeExplicitAutograd) to go into the boxed calling convention. In fact, it is not even necessary to have Python at all; forcing all dispatcher calls through the boxed calling convention is sufficient to trigger the ASAN error. I am guessing what is happening is that when we put the Tensors onto the stack, we avoid refcount bump (because the invariant is that input tensors will stay live for the entirety of the function). When you edit the mutable reference, this invariant is broken (the tensor is freed); ordinarily this is fine because the reference also got updated, but if there is a copy of the Tensor ref on the stack, that reference doesn't get updated, and thus you have the problem.

If you like, we can separate out the "avoid Tensor& modification" into its own patch. But the new code is clearly more correct than the old code.

**Comment by ezyang** [comment]
The proposed alternate solution still overwrites mutable `Tensor&`, which is not allowed. I don't see why it is an improvement.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: Yes, I would like it to be separated into its own patch so that changes to the torch.linalg part of the project are clear and well-tracked.

**Comment by IvanYashchuk** [comment]
> The proposed alternate solution still overwrites mutable Tensor&, which is not allowed. I don't see why it is an improvement.

An improvement would be to handle it in a separate PR with fix-specific tests. ghstack was created for this use case, right?

**Comment by Chillee** [comment]
Ok, moved into separate PR: https://github.com/pytorch/pytorch/pull/85357

**Comment by IvanYashchuk** [comment]
`enable_python_dispatcher` enables a boxed calling convention. Is there any other way to enable this?

A variant of what's in https://github.com/pytorch/pytorch/pull/85239#pullrequestreview-1113249053 is good enough for a test (the result should match with and without boxed calling convention enabled).
